### ISSUE TITLE
Python regression: sparseNdArray.read_sparse_tensor slicing does not work

### ISSUE THREAD

**Issue by bkmartinjr** [single-cell-data/TileDB-SOMA#479]
Recent regression likely related to C++ reader integration (#447):  SparseNdArray.read_as_tensor no longer slices correctly.  If coordinates (ids) are specified, it returns *all* rather than the requested slice. `read_as_pandas_all()` still works.

Example:  I have a large 2D matrix.  Using `read_as_pandas_all` with a slice produces a different result than `read_sparse_tensor`.  They should be identical (in value)

```
In [21]: census['human'].ms['RNA'].X['raw'].read_as_pandas_all((0, slice(None)))
Out[21]: 
      soma_dim_0  soma_dim_1  soma_data
0              0          13        2.0
1              0          18        1.0
2              0          21        1.0
3              0          24        1.0
4              0          29        4.0
...          ...         ...        ...
4974           0       23715        7.0
4975           0       23719        1.0
4976           0       23724        1.0
4977           0       23729        3.0
4978           0       23745        1.0

[4979 rows x 3 columns]

In [22]: for i in census['human'].ms['RNA'].X['raw'].read_sparse_tensor((0, slice(None))): print(i)
<pyarrow.SparseCOOTensor>
type: float
shape: (176, 23747)
```

The (sliced) result above seems to be identical to the entire array, eg,

```
In [23]: for i in census['human'].ms['RNA'].X['raw'].read_sparse_tensor((slice(None), slice(None))): print(i)
<pyarrow.SparseCOOTensor>
type: float
shape: (176, 23747)
```

**Comment by johnkerl** [comment]
# Debugging

@bkmartinjr interesting finding -- I spent a while debugging this today and it is (surprisingly) _not_ a regression -- the bug was there before.

Specifically I checked out commits `ae1554c800c98aca721b7632d309cac797b27230` and `15185875f9831a9164baa91052ac9e387feb56bd` which were before and after #447. 

This script https://gist.github.com/johnkerl/f1331171e1f91fc7eb7312605038cc1f has the same output on the before and after commits, which is

```
================================================================
   soma_dim_0  soma_dim_1  soma_data
0           0           0   2.424371

================================================================
<pyarrow.SparseCOOTensor>
type: float
shape: (65662, 24857)
```

and the bug is [here](https://github.com/single-cell-data/TileDB-SOMA/blob/ae1554c800c98aca721b7632d309cac797b27230/apis/python/src/tiledbsoma/soma_sparse_nd_array.py#L210) on the before commit and [here](https://github.com/single-cell-data/TileDB-SOMA/blob/15185875f9831a9164baa91052ac9e387feb56bd/apis/python/src/tiledbsoma/soma_sparse_nd_array.py#L207) on the after commit.

Another script which zooms in in more detail is [here](https://gist.github.com/johnkerl/121140886ec0274ba472e923d7447174) with output (again same on both branches)

```
COORDS (0, 0)

COORDS 2 (0, 0)

ARROW TABLE
pyarrow.Table
soma_dim_0: int64
soma_dim_1: int64
soma_data: float
----
soma_dim_0: [[0]]
soma_dim_1: [[0]]
soma_data: [[2.4243715]]

COO_DATA
<class 'numpy.ndarray'>
[2.4243715]

N = 0

C = [0]

N = 1

C = [0]

COO_COORDS
<class 'numpy.ndarray'>
(1, 2)
[[0 0]]

YIELD
<pyarrow.SparseCOOTensor>
type: float
shape: (65662, 24857)
```

where the data is right-sized all the way down until we apply the `shape=shape` in

```
pa.SparseCOOTensor.from_numpy(coo_data, coo_coords, shape=shape)
```

# The fix(es)

A couple weeks ago, during our staff-up push, as an organzation we decided to use a division of labor -- bringing in a few more people to create a language-specific spec and align that to the language-independent spec, and to get exhaustive coverage of the implementation with respect to the specs.

Respecting that division of labor, and valuing our new team members, I'll fix this (not new) bug on this PR -- but will not create exhaustive unit-test coverage for the `SparseNdArray` class (or any other) on this PR.

**Comment by bkmartinjr** [comment]
Super odd - I have code that _should_ have failed with this result (but did not fail).   So I assumed that it was a regression.  Apologies!   I will need to further scrutinize my code to figure out why it did not fail given the longer term bug.

If you need me to time travel back and isolate the actual commit, let me know.

**Comment by johnkerl** [comment]
☝️ Specifically I checked out commits ae1554c800c98aca721b7632d309cac797b27230 and 15185875f9831a9164baa91052ac9e387feb56bd which were before and after https://github.com/single-cell-data/TileDB-SOMA/pull/447 -- if you want to take a look

**Comment by johnkerl** [comment]
In any case -- this was a _fun_ debug!! -- regardless it needs fixing. :)

**Comment by johnkerl** [comment]
I think I'll need to create a utility function to take args like `( slice(1,2), slice(None) )` and map those to a new shape -- in this example

* `(1,2)` -> length 2 in dim0
* `slice(None)` -> `A.shape[1]` in dim1

**Comment by johnkerl** [comment]
@bkmartinjr for me the key part of doing that is whether we require _in each dimension_ the coordinate be an int, or a slice (including the none-slice), or an array -- if so, it'll be trivial to compute the new shape from the old shape.

If _in each dimension_ we can have a combination of things, then it gets more complex. In the spec I see

> slice - per-dimension slice, expressed as a scalar, a range, an Arrow array or chunked array of scalar, or a list of both.

which _I think_ permits the latter (more complex) case.

**Comment by johnkerl** [comment]
See also for reference https://gist.github.com/johnkerl/953602d86b6cd1aaf8fecbcc83d21209

**Comment by bkmartinjr** [comment]
Ideally, we would use semantics that are similar to numpy or TileDB-Py, which I understand as:
* a sequence of coordinates is accepted
* sequence length must be at least one and <= number of dimensions. If less than number of dimensions, then "slice(None)" is assumed for the missing dimensions.
* per-dimension, explicitly specified coordinates can be one of:  a value, a list/ndarray/paarray/etc of values, a slice, etc.

So, for a dimension 3 array,  `([1,2,3], )` is equivalent to `([1,2,3], slice(None), slice(None))`

Did I miss any corner cases?

**Comment by bkmartinjr** [comment]
@johnkerl - an related issue I found while testing this:  if you specify an empty list/array for a dimension, it returns "all".  I expected it would return _none_, as this is what Numpy does.  TileDB-Py on the other hand treats an empty coordinate list as _all_.  Confusing.

Numpy-style:
```
In [26]: np.arange(10, dtype=np.float32)[[1]]
Out[26]: array([1.], dtype=float32)

In [27]: np.arange(10, dtype=np.float32)[[]]
Out[27]: array([], dtype=float32)
```

Our current implementation (SOMA and TileDB-Py both do this):
```
In [18]: census['human'].ms['RNA'].X['raw'].read_as_pandas_all([[],slice(None)])
Out[18]: 
        soma_dim_0  soma_dim_1  soma_data
0                0          13        2.0
1                0          18        1.0
2                0          21        1.0
3                0          24        1.0
4                0          29        4.0
...            ...         ...        ...
670301         175       23724        1.0
670302         175       23726        1.0
670303         175       23727        1.0
670304         175       23729        1.0
670305         175       23744        2.0

[670306 rows x 3 columns]
```

I'm not sure which is right.  Personally, I think the NumPy model makes much more sense as it prevents the need for higher level error checking on joins.

**Comment by johnkerl** [comment]
cc @mlin @thetorpedodog

**Comment by johnkerl** [comment]
@bkmartinjr re

> which I understand as:
> ...

I think _for this PR_ I'll do precisely that bullet list, and `raise` on anything else encountered -- and as follow-on we can rely on @thetorpedodog's spec-work & @mlin's unit-test work to revisit/resurface anything beyond that

**Comment by thetorpedodog** [comment]
> Ideally, we would use semantics that are similar to numpy or TileDB-Py, which I understand as:
> 
> * a sequence of coordinates is accepted
> * sequence length must be at least one and <= number of dimensions.
> * If the sequence contains missing coordinates (length less than number of dimensions), then "slice(None)" is assumed for the missing dimensions.
> * per-dimension, explicitly specified coordinates can be one of:  a value, a list/ndarray/paarray/etc of values, a slice, etc.
> 
> So, for a dimension 3 array, `([1,2,3], )` is equivalent to `([1,2,3], slice(None), slice(None))`
> 
> Did I miss any corner cases?
> 
> CC: @aaronwolen @thetorpedodog - in case you guys have any input on Pythonic conventions here.

This overall seems about right with regards to the way TileDB accepts coordinates. I know that there is some weirdness in the way that the APIs there deal with tuples as opposed to lists and how they are interpreted in terms of dimensions/ranges (like `some_arr[a, b, c]` is different from `some_arr[[a, b, c]]`?) but it has been a while since I dealt with that specific part of things. I would want to make sure we don’t effectively codify that exact behavior (since it seems like a very easy rake to step on).

**Comment by bkmartinjr** [comment]
The one thing I definitely need for my use cases is the ability to call with multiple dimensions specified in a sequence (tuple usually, and this seems like the minimum necessary type with list as optional/nice-to-have).

Eg., `(dim1coord, dim2coord, ...)`   Concrete example:   `([0,3,4], slice(2,10))`

**Comment by johnkerl** [comment]
@bkmartinjr 💯
